### ISSUE TITLE
invalidateAfter isn't very efficient

### ISSUE THREAD

**Issue by ElliottLMz** [ElliottLMz/Cachet#1]
[NovusTheory pointed this out on the DevForum](https://devforum.roblox.com/t/cachet-an-open-source-module-for-robust-caching/421770/6?u=elliottlmz). The method the system uses to invalidate cache entries on queue is very inefficient and could potentially cause a performance hit on a low-end client, or if enough cache entries are pending invalidation.

Delay could work however I have heard that it may be inefficient in some cases. Can anyone provide more details on it?

NovusTheory suggested that Cachet follows what Redis does and polls to check for invalid keys, however I don't believe that would be very efficeint.

It would require Cachet to keep a global record of caches, and then to (at least every second to ensure things work as intended) run through all caches and their keys to see if they are due/overdue invalidation.

**Comment by NovusTheory** [comment]
[Here's the relevant active expiration code for Redis](https://github.com/antirez/redis/blob/a92921da135e38eedd89138e15fe9fd1ffdd9b48/src/expire.c#L98) if you want to know more about how Redis does it.

I wouldn't consider it inefficient to do given you attempt what Redis does and how it's used by a lot of major service providers.

**Comment by ElliottLMz** [comment]
> I wouldn't consider it inefficient to do given you attempt what Redis does and how it's used by a lot of major service providers.

This may be efficient in C, however I can't think of any fast ways to do this in Lua without creating a sort of 'tasks' tracker, which I'd like to avoid doing.

**Comment by NovusTheory** [comment]
All Redis is doing is running some function every so often and ensuring it doesn't run for a long period of time. It's not doing anything complex.

**Comment by ElliottLMz** [comment]
I've been looking into this and feel like this is a non-issue. Cachet uses courontines which internally work similar to `delay`, but better. Delay can, at random, simply refuse to work- this makes it unsuitable for use. I did consider tracking invalidations, but I feel like it would be too inefficient to do- and would require a major refactor of Cachet's codebase, which defeats it's aim of being a relatively simple module. 

If anyone does know a more efficient method for the 'invalidateAfter' feature, please [open an issue](https://github.com/ElliottLMz/Cachet/issues/new). In the meantime, I'll be closing this out as a wontfix.

**Comment by howmanysmall** [comment]
A bit late, but you could use a function like this instead:

```lua
--[[**
	Schedules a function to be executed after DelayTime seconds have passed, without yielding the current thread. This function allows multiple Lua threads to be executed in parallel from the same stack. Unlike regular delay, this does not use the legacy scheduler and also allows passing arguments.
	@param [t:numberMin<0>] DelayTime The amount of time before the function will be executed.
	@param [t:callback] Function The function you are executing.
	@param [variant] ... The optional arguments you can pass that the function will execute with.
	@returns [void]
**--]]
function Scheduler.Delay(DelayTime, Function, ...)
	local Length = select("#", ...)
	if Length > 0 then
		local Arguments = {...}
		local ExecuteTime = tick() + DelayTime
		local Connection

		Connection = Heartbeat:Connect(function()
			if tick() >= ExecuteTime then
				Connection:Disconnect()
				Function(table.unpack(Arguments, 1, Length))
			end
		end)

		return Connection
	else
		local ExecuteTime = tick() + DelayTime
		local Connection

		Connection = Heartbeat:Connect(function()
			if tick() >= ExecuteTime then
				Connection:Disconnect()
				Function()
			end
		end)

		return Connection
	end
end
```
